[DAGCombiner] Add support for scalarising extracts of a vector setcc

david-arm opened the pull request:
For IR like this:

%icmp = icmp ult <4 x i32> %a, splat (i32 5)
%res = extractelement <4 x i1> %icmp, i32 1

where there is only one use of %icmp we can take a similar approach
to what we already do for binary ops such add, sub, etc. and convert
this into

%ext = extractelement <4 x i32> %a, i32 1
%res = icmp ult i32 %ext, 5

For AArch64 targets at least the scalar boolean result will almost
certainly need to be in a GPR anyway, since it will probably be
used by branches for control flow. I've tried to reuse existing code
in scalarizeExtractedBinop to also work for setcc.

NOTE: The optimisations don't apply for tests such as
extract_icmp_v4i32_splat_rhs in the file

CodeGen/AArch64/extract-vector-cmp.ll

because scalarizeExtractedBinOp only works if one of the input
operands is a constant.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22746,16 +22746,21 @@ SDValue DAGCombiner::scalarizeExtractedVectorLoad(SDNode *EVE, EVT InVecVT,
   auto *IndexC = dyn_cast<ConstantSDNode>(Index);
-  if (!IndexC || !TLI.isBinOp(Vec.getOpcode()) || !Vec.hasOneUse() ||
+  unsigned Opc = Vec.getOpcode();
+  if (!IndexC || (!TLI.isBinOp(Opc) && Opc != ISD::SETCC) || !Vec.hasOneUse() ||
```
```suggestion
  if (!IndexC || !Vec.hasOneUse() || (Opc != ISD::SETCC && !TLI.isBinOp(Opc)) || 
```

david-arm replied:
Done

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22766,19 +22771,25 @@ static SDValue scalarizeExtractedBinop(SDNode *ExtElt, SelectionDAG &DAG,
-    SDValue Ext0 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, VT, Op0, Index);
-    SDValue Ext1 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, VT, Op1, Index);
-    return DAG.getNode(Vec.getOpcode(), DL, VT, Ext0, Ext1);
+  if (!isAnyConstantBuildVector(Op0, true) &&
+      !ISD::isConstantSplatVector(Op0.getNode(), SplatVal) &&
+      !isAnyConstantBuildVector(Op1, true) &&
+      !ISD::isConstantSplatVector(Op1.getNode(), SplatVal))
+    return SDValue();
```
This is a confusing predicate. It would be better to rewrite in terms of one predicate on each operand

david-arm replied:
I see we already have a confusing number of predicate functions in SelectionDAG and DAGCombiner related to asking similar, but not identical questions about whether an operand is a constant splat/non-constant splat/constant build vector/etc. If it's ok with you I'd prefer not to add one more for now in this patch, but I'm happy to do a follow-on patch that tries to rationalise/refactor some of the existing interfaces. Hopefully that will then allow me to rewrite and simplify this code. What do you think?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22766,19 +22771,25 @@ static SDValue scalarizeExtractedBinop(SDNode *ExtElt, SelectionDAG &DAG,
+    Op1 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, OpVT, Op1, Index);
+    return DAG.getSetCC(DL, ResVT, Op0, Op1,
+                        cast<CondCodeSDNode>(Vec->getOperand(2))->get());
+  } else {
```
No else after return

david-arm replied:
Done

arsenm commented on llvm/test/CodeGen/AArch64/extract-vector-cmp.ll:
```diff
@@ -0,0 +1,225 @@
+  br i1 %3, label %pred.store.if5, label %pred.store.continue6
+
+pred.store.if5:
+  %4 = or disjoint i64 %index, 1
```
Use named values in tests

arsenm commented on llvm/test/CodeGen/AArch64/extract-vector-cmp.ll:
```diff
@@ -0,0 +1,225 @@
+  store i32 1, ptr %5, align 4
+  br label %pred.store.continue6
+
+pred.store.continue6:
```
Function really big for a dag combine test

david-arm replied:
I agree it was probably too big. I was trying to highlight the motivating example from LoopVectorize.cpp and demonstrate the improvement. I've reduced the test to use a <2 x i64> instead to reduce the size. Hopefully this is small enough now!

I've also tried to properly name all the variables.

david-arm committed:
Add tests

* Rewrote the test in dag-combine-concat-vectors.ll to ensure there
is more than one use of the vector comparison.

david-arm committed:
[DAGCombiner] Add support for scalarising extracts of a vector setcc

For IR like this:

  %icmp = icmp ult <4 x i32> %a, splat (i32 5)
  %res = extractelement <4 x i1> %icmp, i32 1

where there is only one use of %icmp we can take a similar approach
to what we already do for binary ops such add, sub, etc. and convert
this into

  %ext = extractelement <4 x i32> %a, i32 1
  %res = icmp ult i32 %ext, 5

For AArch64 targets at least the scalar boolean result will almost
certainly need to be in a GPR anyway, since it will probably be
used by branches for control flow. I've tried to reuse existing code
in scalarizeExtractedBinop to also work for setcc.

NOTE: The optimisations don't apply for tests such as
extract_icmp_v4i32_splat_rhs in the file

CodeGen/AArch64/extract-vector-cmp.ll

because scalarizeExtractedBinOp only works if one of the input
operands is a constant.

david-arm committed:
Address review comment

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22766,19 +22771,24 @@ static SDValue scalarizeExtractedBinop(SDNode *ExtElt, SelectionDAG &DAG,
+  // extractelt (op X, C), IndexC --> op (extractelt X, IndexC), C'
+  // extractelt (op C, X), IndexC --> op C', (extractelt X, IndexC)
+  if (Opc == ISD::SETCC) {
+    EVT OpVT = Op0->getValueType(0).getVectorElementType();
```
Done!

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -22766,19 +22772,24 @@ static SDValue scalarizeExtractedBinop(SDNode *ExtElt, SelectionDAG &DAG,
+  // extractelt (op X, C), IndexC --> op (extractelt X, IndexC), C'
+  // extractelt (op C, X), IndexC --> op C', (extractelt X, IndexC)
+  if (Opc == ISD::SETCC) {
+    EVT OpVT = Op0->getValueType(0).getVectorElementType();
```
```suggestion
    EVT OpVT = Op0.getValueType().getVectorElementType();
```

david-arm committed:
Address review comment

vitalybuka commented:
This patch breaks https://lab.llvm.org/buildbot/#/builders/55/builds/3959

vitalybuka commented:
@david-arm Please acknowledge that you noticed the revert.

Please let me know if you need help with reproducer, but I guess it's quite straight forward either using https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild or copy-pasting cmake commands from the bot.

david-arm commented:
> @david-arm Please acknowledge that you noticed the revert.
> 
> Please let me know if you need help with reproducer, but I guess it's quite straight forward either using https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild or copy-pasting cmake commands from the bot.

Yep, I'm taking a look now! The error does look suspiciously like it's caused by this patch.
